wandb/sdk/artifacts/storage_handlers/fsspec_file_handler.py

```diff
@@ -0,0 +1,146 @@
+
+
+class FsspecFileHandler(StorageHandler):
+    """Handles a variety of different storage solutions."""
+
+    def __init__(self, scheme: Optional[str] = None) -> None:
+        """Track files or directories on a variety of filesystem.
+
+        For now, this handler supports oss (Alibaba Object Storage System)
+        A list of all available options can be found under the following link:
+        https://github.com/fsspec/filesystem_spec/blob/master/fsspec/registry.py
+        """
+        self._scheme = scheme or "oss"
+        self._cache = get_artifact_file_cache()
+
+    def can_handle(self, parsed_url: "ParseResult") -> bool:
+        return parsed_url.scheme in self._scheme
+
+    def load_path(
+        self,
+        manifest_entry: ArtifactManifestEntry,
+        local: bool = False,
+    ) -> Union[URIStr, FilePathStr]:
+        """Load the file in the specified artifact given its corresponding entry.
+
+        Download the referenced artifact; return the path to the caller.
+
+        Arguments:
+            manifest_entry (ArtifactManifestEntry): The index entry to load
+
+        Returns:
+            (os.PathLike): A path to the file represented by `index_entry`
+        """
+        if not local:
+            assert manifest_entry.ref is not None
+            return manifest_entry.ref
+
+        if manifest_entry.ref is None:
+            raise ValueError(f"Cannot add path with no ref: {manifest_entry.path}")
+        fs, fs_path = fsspec.core.url_to_fs(str(manifest_entry.ref))
+
+        if not fs.exists(fs_path):
+            raise ValueError(
+                "fsspec file reference: Failed to find file at path %s" % fs_path
+            )
+
+        path, hit, cache_open = self._cache.check_md5_obj_path(
+            B64MD5(manifest_entry.digest),  # TODO(spencerpearson): unsafe cast
+            manifest_entry.size if manifest_entry.size is not None else 0,
+        )
+        if hit:
+            return path
+
+        md5 = fs.checksum(fs_path)
+        if md5 != manifest_entry.digest:
+            raise ValueError(
+                f"fsspec file reference: Digest mismatch for path {manifest_entry.path}: expected {manifest_entry.digest} but found {md5}"
+            )
+
+        filesystem.mkdir_exists_ok(os.path.dirname(path))
+        fs.get_file(fs_path, path)
+
+        return path
+
+    def store_path(
+        self,
+        artifact: "Artifact",
+        path: Union[URIStr, FilePathStr],
+        name: Optional[StrPath] = None,
+        checksum: bool = True,
+        max_objects: Optional[int] = None,
+    ) -> Sequence[ArtifactManifestEntry]:
+        max_objects = max_objects or DEFAULT_MAX_OBJECTS
+        # We have a single file or directory
+        entries = []
+        fs, fs_path = fsspec.core.url_to_fs(path)
+
+        def md5(path: str) -> Any:
+            return (
+                fs.checksum(path)
+                if checksum
+                else md5_string(str(fs.info(path)["size"]))
```

Comment:
We use the file digest to know if the file content changed, and thus we should create a new artifact version. People often have files that never change. In this case they can set `checksum=False` and we use the file URL as its digest. Admittedly, this should be explained better [here](https://github.com/wandb/wandb/blob/6d7ca08e1b5b2d2f14c2b8e16cd00198520cfcb6/wandb/sdk/artifacts/storage_handler.py#L53).

Using file size as its digest is a reasonable heuristic. It won't work properly in the use case where files never change though. The following code will only create one artifact if `first_file.txt` and `second_file.txt` are different but happen to have the same size:
```
artifact = wandb.Artifact("my_artifact", type="my_type")
artifact.add_reference("oss://my_bucket/first_file.txt", name="my_file", checksum=False)
artifact.save()

artifact = wandb.Artifact("my_artifact", type="my_type")
artifact.add_reference("oss://my_bucket/second_file.txt", name="my_file", checksum=False)
artifact.save()
```

Reply:
Due to the fact that I added this at the top: 
```
        if not checksum:
            entry_path = name or (key if key != "" else bucket)
            return [ArtifactManifestEntry(path=entry_path, ref=path, digest=path)]
```
This code will actually never be reached if `checksum=False` correct?
```
        return (
            fs.checksum(path)
            if checksum
            else md5_string(str(fs.info(path)["size"]))
```
So currently, this `md5_string(str(fs.info(path)["size"]))` will never be executed / reached...